src/msix/pack/AppxBundleWriter.cpp

```diff
@@ -0,0 +1,280 @@
+        ThrowHrIfFailed(stream->Seek(start, StreamBase::Reference::START, nullptr));
+        std::uint64_t uncompressedSize = static_cast<std::uint64_t>(end.QuadPart);
+
+        // Add file to block map.
+        if (addToBlockMap)
+        {
+            m_blockMapWriter.AddFile(name, uncompressedSize, fileInfo.first);
+        }
+
+        auto& zipFileStream = fileInfo.second;
+
+        std::uint64_t bytesToRead = uncompressedSize;
+        std::uint32_t crc = 0;
+        while (bytesToRead > 0)
+        {
+            // Calculate the size of the next block to add
+            std::uint32_t blockSize = (bytesToRead > DefaultBlockSize) ? DefaultBlockSize : static_cast<std::uint32_t>(bytesToRead);
+            bytesToRead -= blockSize;
+
+            // read block from stream
+            std::vector<std::uint8_t> block;
+            block.resize(blockSize);
+            ULONG bytesRead;
+            ThrowHrIfFailed(stream->Read(static_cast<void*>(block.data()), static_cast<ULONG>(blockSize), &bytesRead));
+            ThrowErrorIfNot(Error::FileRead, (static_cast<ULONG>(blockSize) == bytesRead), "Read stream file failed");
+            crc = crc32(crc, block.data(), static_cast<uInt>(block.size()));
+
+            // Write block and compress if needed
+            ULONG bytesWritten = 0;
+            ThrowHrIfFailed(zipFileStream->Write(block.data(), static_cast<ULONG>(block.size()), &bytesWritten));
+
+            // Add block to blockmap
+            if (addToBlockMap)
+            {
+                m_blockMapWriter.AddBlock(block, bytesWritten, toCompress);
+            }
+
+        }
+
+        if (toCompress)
+        {
+            // Put the stream termination on
+            std::vector<std::uint8_t> buffer;
+            ULONG bytesWritten = 0;
+            ThrowHrIfFailed(zipFileStream->Write(buffer.data(), static_cast<ULONG>(buffer.size()), &bytesWritten));
+        }
+
+        // Close File element
+        if (addToBlockMap)
+        {
+            m_blockMapWriter.CloseFile();
+        }
+
+        // This could be the compressed or uncompressed size
+        auto streamSize = zipFileStream.As<IStreamInternal>()->GetSize();
+        m_zipWriter->EndFile(crc, streamSize, uncompressedSize, true);
+    }
+
+    void AppxBundleWriter::ValidateCompressionOption(APPX_COMPRESSION_OPTION compressionOpt)
+    {
+        bool result = ((compressionOpt == APPX_COMPRESSION_OPTION_NONE) ||
+                       (compressionOpt == APPX_COMPRESSION_OPTION_NORMAL) ||
+                       (compressionOpt == APPX_COMPRESSION_OPTION_MAXIMUM) ||
+                       (compressionOpt == APPX_COMPRESSION_OPTION_FAST) ||
+                       (compressionOpt == APPX_COMPRESSION_OPTION_SUPERFAST));
+        ThrowErrorIfNot(Error::InvalidParameter, result, "Invalid compression option.");
+    }
+
+    // Common close functionality for pack and bundle.
+    // Adds AppxBlockMap.xml, [Content_Types].xml and closes the zip object.

```

Comment:
You can remove this. My idea was to use AppxPackageWriter as both the package and the bundle writer, but since its now separate this comment shouldn't go here

Reply:
Removed CloseInternal method and moved it to ::Close in the AppxBundleWriter.cpp file.